api/src/main/java/jakarta/mail/util/SharedFileInputStream.java

```diff
@@ -474,19 +459,15 @@ public long getPosition() {
      */
     @Override
     public synchronized InputStream newStream(long start, long end) {
-        try {
-            if (in == null)
-                throw new RuntimeException("Stream closed");
-            if (start < 0)
-                throw new IllegalArgumentException("start < 0");
-            if (end == -1)
-                end = datalen;
-
-            return new SharedFileInputStream(sf,
-                    this.start + start, end - start, bufsize);
-        } finally {
-            Objects.requireNonNull(this); //TODO: replace with Reference.reachabilityFence
```

Comment:
I am analyzing this class deeply. At the end there is the same instance of `RandomAccessFile` for the different `SharedFileInputStream` amount of instances. Only the latest one to be closed will really close the `RandomAccessFile`.

I understand that I need to make sure the counters are correctly updated (with synchronized, reentrant locks or atomic integers), to avoid concurrency issues that could lead to an unclosed `RandomAccessFile`. But I don't understand your example of `goes: 0,1a,0,1b instead of 0,1a,2,1b`.

Maybe you can illustrate it with one example. I am doing some tests related to garbage collection [here](https://github.com/jbescos/mail/blob/0540a360acd815cf963e284f204cab5612bd05cf/api/src/test/java/jakarta/mail/util/SharedFileInputStreamTest.java#L97), so I can try to reproduce your example.

Reply:
https://github.com/jakartaee/mail-api/blob/bf2bfc18c0d2b9b3f57417203c2a462a3a94a7de/api/src/test/java/jakarta/mail/util/SharedFileInputStreamTest.java#L38

In testing you have to be careful not to pin an instance by using a local ref or you'll fool yourself in to thinking 'this' can't be subject to G.C.

With this change that test case should be updated so that the root class is monitored by a cleaner or is a subclass with finallizer.

I can write up a change.

Reply:
Here is the summary of the deep dive on this topic:
https://mail.openjdk.org/pipermail/core-libs-dev/2018-February/051312.html

The whole thread is there if you want to dig further.

Reply:
Just to double check that I understand this, do you mean that in the next line:
`new SharedFileInputStream(file).newStream(0, -1)`
That `new SharedFileInputStream(file)` could be GCed meanwhile `newStream(0, -1)` is happening?.

In that case, there could be a problem in the current implementation, because `#finalize` could be invoked (this currently executes `#close`) and the sharedfile closed meanwhile `newStream(0, -1)` is happening. If we remove the `#finalize` this cannot happen.

This is a tricky thing, I am sorry for all the questions but I want to understand it correctly.

Reply:
I created the next test, and I cannot reproduce the issue removing Objects.requireNonNull(this);. The method `#finalize` is not invoked meanwhile `#newStream` is in progress:

```
    @Test
    public void testGC() throws Exception {
        File file = new File(SharedFileInputStreamTest.class.getResource("/jakarta/mail/util/sharedinputstream.txt").toURI());
        try (SharedFileInputStream in = (SharedFileInputStream) new SubclassFinalize(file).newStream(0, 6)) {
            assertEquals("test0\n", new String(in.readAllBytes()));
        }
    }

    private static class SubclassFinalize extends SharedFileInputStream {

        private final CountDownLatch latch = new CountDownLatch(1);

        public SubclassFinalize(File file) throws IOException {
            super(file);
        }

        @Override
        public synchronized InputStream newStream(long start, long end) {
            try {
                // Wait till root dependency gets collected
                latch.await(10, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                // Ignore
            }
            return super.newStream(start, end);
        }

        @Override
        protected void finalize() throws Throwable {
            super.close();
            latch.countDown();
        }
    }
```

But anyway, maybe in other java versions it could happen. I can keep the Objects.requireNonNull(this);

Reply:
This test seems to be passing without the `Objects.requireNonNull(this)`:

```
    @Test
    public void testGC() throws Exception {
        File file = new File(SharedFileInputStreamTest.class.getResource("/jakarta/mail/util/sharedinputstream.txt").toURI());
        try (SharedFileInputStream in = (SharedFileInputStream) new SubclassFinalize(file).newStream(0, 6)) {
            assertEquals("test0\n", new String(in.readAllBytes()));
        }
    }

    private static class SubclassFinalize extends SharedFileInputStream {

        private final CountDownLatch latch = new CountDownLatch(1);

        public SubclassFinalize(File file) throws IOException {
            super(file);
        }

        private SubclassFinalize(SharedFile sf, long start, long len, int bufsize, CountDownLatch latch) {
            super(sf, start, len, bufsize);
            try {
                latch.await(10, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                // Ignore
            }
        }

        @Override
        public synchronized InputStream newStream(long start, long end) {
            try {
                SharedFile sf = getSharedFile(this);
                return new SubclassFinalize(sf,
                        this.start + start, end - start, bufsize, latch);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }

        @Override
        protected void finalize() throws Throwable {
            super.close();
            latch.countDown();
        }
    }
```

Maybe is my java version. We can keep `Objects.requireNonNull(this)` just in case.

Reply:
The getSharedFile(this) is most likely keeping instance alive.  If we are not supporting JDK8 then the code should use try/finally Reference.reachabilityFence(this).

Reply:
I added the method `getSharedFile(this)` so you can see it.

Reply:
Sounds good.  The synchronized keyword makes it so `newStream` is safe from early G.C.  Might want to either add fence back there or at least add a comment warning those that are going to remove synchronized keyword.

Reply:
Right, I forgot that, thank you. It is in place now.